libs/math/tests/unit/tensor/tensor_operations.cpp

```diff
@@ -21,21 +21,85 @@
 #include "test_types.hpp"
 #include "vectorise/fixed_point/fixed_point.hpp"
 
-namespace fetch {
-namespace math {
-namespace test {
+namespace {
+
+using SizeType = fetch::math::SizeType;
 
 template <typename T>
 class TensorOperationsTest : public ::testing::Test
 {
 };
 
-TYPED_TEST_CASE(TensorOperationsTest, FloatIntAndUIntTypes);
+TYPED_TEST_CASE(TensorOperationsTest, fetch::math::test::FloatIntAndUIntTypes);
 
-TYPED_TEST(TensorOperationsTest, inline_add_test)
+template <typename TypeParam>
+void ReshapeTest(std::vector<SizeType> const &initial_shape,
+                 std::vector<SizeType> const &final_shape)
+{
+  fetch::math::Tensor<TypeParam> t1(initial_shape);
+  fetch::math::Tensor<TypeParam> t2(final_shape);
+  t1.Reshape(final_shape);
+  EXPECT_TRUE(t1.shape() == t2.shape());
+}
+
+TYPED_TEST(TensorOperationsTest, two_dim_reshape_test)
 {
-  using SizeType = fetch::math::SizeType;
+  ReshapeTest<TypeParam>({2, 3}, {6, 1});
+  ReshapeTest<TypeParam>({4, 4}, {2, 8});
+  ReshapeTest<TypeParam>({1, 2}, {2, 1});
+  ReshapeTest<TypeParam>({0, 0}, {0, 0});
+}
+
+TYPED_TEST(TensorOperationsTest, three_and_two_dim_reshape_test)
+{
+  ReshapeTest<TypeParam>({2, 3, 1}, {6, 1});
+  ReshapeTest<TypeParam>({2, 1, 3}, {6, 1});
+  ReshapeTest<TypeParam>({1, 2, 3}, {6, 1});
+  ReshapeTest<TypeParam>({3, 2, 1}, {1, 6});
+
+  ReshapeTest<TypeParam>({6, 1}, {2, 3, 1});
+  ReshapeTest<TypeParam>({6, 1}, {2, 1, 3});
+  ReshapeTest<TypeParam>({6, 1}, {1, 2, 3});
+  ReshapeTest<TypeParam>({1, 6}, {3, 2, 1});
+}
 
+TYPED_TEST(TensorOperationsTest, three_dim_reshape_test)
+{
+  ReshapeTest<TypeParam>({2, 3, 1}, {2, 1, 3});
+  ReshapeTest<TypeParam>({2, 1, 3}, {3, 2, 1});
+  ReshapeTest<TypeParam>({1, 2, 3}, {2, 1, 3});
+  ReshapeTest<TypeParam>({3, 2, 1}, {1, 2, 3});
+  ReshapeTest<TypeParam>({1, 2, 3}, {1, 2, 3});
+}
+
+TYPED_TEST(TensorOperationsTest, four_dim_to_two_dim_reshape_test)
+{
+  ReshapeTest<TypeParam>({2, 3, 1, 1}, {6, 1});
+  ReshapeTest<TypeParam>({2, 1, 3, 1}, {6, 1});
+  ReshapeTest<TypeParam>({2, 1, 1, 3}, {6, 1});
+  ReshapeTest<TypeParam>({2, 3, 4, 5}, {12, 10});
+
+  ReshapeTest<TypeParam>({6, 1}, {2, 3, 1, 1});
+  ReshapeTest<TypeParam>({6, 1}, {2, 1, 3, 1});
+  ReshapeTest<TypeParam>({6, 1}, {2, 1, 1, 3});
+  ReshapeTest<TypeParam>({12, 10}, {2, 3, 4, 5});
+}
+
+TYPED_TEST(TensorOperationsTest, four_and_three_dim_reshape_test)
+{
+  ReshapeTest<TypeParam>({2, 3, 1, 1}, {6, 1, 1});
+  ReshapeTest<TypeParam>({2, 1, 3, 1}, {6, 1, 1});
+  ReshapeTest<TypeParam>({2, 1, 1, 3}, {2, 1, 3});
+  ReshapeTest<TypeParam>({2, 3, 4, 5}, {12, 2, 5});
+
+  ReshapeTest<TypeParam>({6, 1, 1}, {2, 3, 1, 1});
+  ReshapeTest<TypeParam>({2, 1, 3}, {2, 1, 3, 1});
+  ReshapeTest<TypeParam>({1, 1, 6}, {2, 1, 1, 3});
+  ReshapeTest<TypeParam>({3, 4, 10}, {2, 3, 4, 5});
+}
+
```

Comment:
What about negative cases, when reshape should fail/throw ?

Reply:
The parameters are vectors of SizeTypes, so presumably it would give a compilation error.